execution_engine/src/storage/global_state/scratch.rs

```diff
@@ -0,0 +1,498 @@
+        empty_root_hash: Digest,
+    ) -> Self {
+        ScratchGlobalState {
+            cache: Arc::new(RwLock::new(Cache::new())),
+            environment,
+            trie_store,
+            empty_root_hash,
+        }
+    }
+
+    /// Consume self and return inner cache.
+    pub fn into_inner(self) -> HashMap<Key, StoredValue> {
+        let cache = mem::replace(&mut *self.cache.write().unwrap(), Cache::new());
+        cache.into_inner()
+    }
+}
+
+impl StateReader<Key, StoredValue> for ScratchGlobalStateView {
+    type Error = error::Error;
+
+    fn read(
+        &self,
+        correlation_id: CorrelationId,
+        key: &Key,
+    ) -> Result<Option<StoredValue>, Self::Error> {
+        if let Some(value) = self.cache.read().unwrap().get(key) {
+            return Ok(Some(value.clone()));
+        }
+        let txn = self.environment.create_read_txn()?;
+        let ret = match read::<Key, StoredValue, lmdb::RoTransaction, LmdbTrieStore, Self::Error>(
+            correlation_id,
+            &txn,
+            self.store.deref(),
+            &self.root_hash,
+            key,
+        )? {
+            ReadResult::Found(value) => {
+                self.cache.write().unwrap().insert(*key, value.clone());
+                Some(value)
+            }
+            ReadResult::NotFound => None,
+            ReadResult::RootNotFound => panic!("ScratchGlobalState has invalid root"),
+        };
+        txn.commit()?;
+        Ok(ret)
+    }
+
+    fn read_with_proof(
+        &self,
+        correlation_id: CorrelationId,
+        key: &Key,
+    ) -> Result<Option<TrieMerkleProof<Key, StoredValue>>, Self::Error> {
+        let txn = self.environment.create_read_txn()?;
+        let ret = match read_with_proof::<
+            Key,
+            StoredValue,
+            lmdb::RoTransaction,
+            LmdbTrieStore,
+            Self::Error,
+        >(
+            correlation_id,
+            &txn,
+            self.store.deref(),
+            &self.root_hash,
+            key,
+        )? {
+            ReadResult::Found(value) => Some(value),
+            ReadResult::NotFound => None,
+            ReadResult::RootNotFound => panic!("LmdbWithCacheGlobalState has invalid root"),
+        };
+        txn.commit()?;
+        Ok(ret)
+    }
+
+    fn keys_with_prefix(
+        &self,
+        correlation_id: CorrelationId,
+        prefix: &[u8],
+    ) -> Result<Vec<Key>, Self::Error> {
+        let txn = self.environment.create_read_txn()?;
+        let keys_iter = keys_with_prefix::<Key, StoredValue, _, _>(
+            correlation_id,
+            &txn,
+            self.store.deref(),
+            &self.root_hash,
+            prefix,
+        );
+        let mut ret = Vec::new();
+        for result in keys_iter {
+            match result {
+                Ok(key) => ret.push(key),
+                Err(error) => return Err(error),
+            }
+        }
+        txn.commit()?;
+        Ok(ret)
+    }
+}
```

Comment:
I hope we can get rid of this code duplication when we implement the trie-level caching.

Reply:
Isn't trie-level caching obviated by bulk-writing the trie?